### PR TITLE
Active learning

### PR DESCRIPTION
**Pull request by aryamansriram**
### Summary
This branch adds the active learning functionality for the training procedure

### CODE REVIEW

**Comment by aryamansriram** [comment]
[Active Learner experimentation](https://trello.com/c/JfjXGB14/30-active-learner-experimentation)

File: models/binary_relevance_active.py [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score
+from tensorflow.keras.wrappers.scikit_learn import KerasClassifier
+from sklearn.neighbors import KNeighborsClassifier, KNeighborsRegressor
+
+import matplotlib.pyplot as plt
+import tensorflow as tf
+from youtube_audioset import get_recursive_sound_names
+
+from tensorflow.compat.v1.keras.models import Model
+from preprocess_utils import import_dataframes, get_select_vector
```
Comment (nandan-sridhar) [inline models/binary_relevance_active.py:26]:
```suggestion
from .preprocess_utils import import_dataframes, get_select_vector
```

File: models/binary_relevance_active.py [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+#############################################################################
+# Importing dataframes from the function
+#############################################################################
+DF_TRAIN, _ = \
+    import_dataframes(dataframe_file_list=CONFIG_DATA["train"]["inputDataFrames"],
+                      positive_label_filter_arr=POSITIVE_LABELS,
+                      negative_label_filter_arr=NEGATIVE_LABELS,
+                      validation_split=CONFIG_DATA["train"]["validationSplit"])
+
+with open("diff_class_datasets/Datasets/Nandi.pkl", "rb") as f:
```
Comment (nandan-sridhar) [inline models/binary_relevance_active.py:117]: I don't have this pickle file. Can you make the path to the pickle file an argument to the script?

File: models/binary_relevance_active.py [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+plt.savefig("Recall_vs_query.png")
+plt.show()
+
+plt.plot(q_list_active, tnr_active, "blue", label="active_learning")
+plt.plot(q_list_rand, tnr_rand, "red", label="random_sampling")
+plt.xlabel("Query Index")
+plt.ylabel("TNR")
+plt.legend(['active_learning', 'random_sampling'], loc='upper left')
+plt.savefig("TNR_vs_query.png")
+plt.show()
```
Comment (nandan-sridhar) [inline models/binary_relevance_active.py:570]: Please consolidate this into one plot with subplots. Please add a title for each subplot.

File: models/binary_relevance_active.py [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+        labels_pool_arr = np.delete(labels_pool_arr, man_idx, axis=0)
+
+    if strategy == "active":
+        with open("queried_ex.pkl", "wb") as f:
+            pickle.dump(queries, f)
+    plt.plot(query_list[1:], train_acc)
+    plt.xlabel("Query Index")
+    plt.ylabel("Train Loss")
+
+    plt.show()
```
Comment (nandan-sridhar) [inline models/binary_relevance_active.py:524]: Please move this down with all the other plots. Add this as a subplot.

File: models/binary_relevance_active.py [hunk 5]
```diff
@@ -46,6 +46,8 @@
 REQUIRED_NAMED.add_argument('-model_cfg_json', '--model_cfg_json',
                             help='Input json configuration file for label',
                             required=True)
+REQUIRED_NAMED.add_argument("-external_datapath", "--external_datapath", help="Path to external data",
```
Comment (nandan-sridhar) [inline models/binary_relevance_active.py:49]: Please also add description of where this data is used (e.g. unlabeled pool and part of val set)

File: models/binary_relevance_active.py [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+# preprocess the data into required structure
+#############################################################################
+
+
+X_TRAIN = np.array(DF_TRAIN.features.apply(lambda x: x.flatten()).tolist())
+X_TRAIN_STANDARDIZED = X_TRAIN / 255
+X_TEST = DF_TEST.features.apply(lambda x: x.flatten()).tolist()
+
+drop_indices = []
+for ii, i in enumerate(X_TEST):
```
Comment (nandan-sridhar) [inline models/binary_relevance_active.py:180]: Please use better variable names than i or ii

File: models/binary_relevance_active.py [hunk 7]
```diff
@@ -0,0 +1,567 @@
...
+
+def read_pool(pool_path):
+    with open(pool_path, "rb") as f:
+        df = pickle.load(f)
+    # df = pd.read_csv(pool_path)
+    split_wavfiles = []
+    split_labels = []
+    for ii, feat in enumerate(df.features):
+
+        for i in range(0, (len(feat) // 10) * 10, 10):
```
Comment (nandan-sridhar) [inline models/binary_relevance_active.py:283]: Same here. Please use better variable names than i or ii

**Pull request merged**